### PR TITLE
Implement DataFrame.LoadCsvFromString

### PR DESCRIPTION
**Pull request by pgovind**
Fixes https://github.com/dotnet/corefxlab/issues/2984

### CODE REVIEW

File: src/Microsoft.Data.Analysis/DataFrame.IO.cs [hunk 1]
```diff
@@ -174,31 +172,12 @@ private static DataFrameColumn CreateColumn(Type kind, string[] columnNames, int
...
-        /// <param name="header">has a header or not</param>
-        /// <param name="columnNames">column names (can be empty)</param>
-        /// <param name="dataTypes">column types (can be empty)</param>
-        /// <param name="numberOfRowsToRead">number of rows to read not including the header(if present)</param>
-        /// <param name="guessRows">number of rows used to guess types</param>
-        /// <param name="addIndexColumn">add one column with the row index</param>
-        /// <param name="encoding">The character encoding. Defaults to UTF8 if not specified</param>
-        /// <returns><see cref="DataFrame"/></returns>
-        public static DataFrame LoadCsv(Stream csvStream,
+        private static DataFrame ReadCsvLinesIntoDataFrame(IEnumerable<string> lines,
```
Comment (pgovind) [inline src/Microsoft.Data.Analysis/DataFrame.IO.cs:175]: This method is just a refactor of the logic in `LoadCsv` with 1 simple change: The first parameter here is an `IEnumerable<string>`.

File: src/Microsoft.Data.Analysis/DataFrame.IO.cs [hunk 2]
```diff
@@ -214,87 +193,189 @@ private static DataFrameColumn CreateColumn(Type kind, string[] columnNames, int
...
+            }
+
+            if (dataTypes == null && guessRows <= 0)
+            {
+                throw new ArgumentException(string.Format(Strings.ExpectedEitherGuessRowsOrDataTypes, nameof(guessRows), nameof(dataTypes)));
+            }
+
+            using (var streamReader = new StreamReader(csvStream, encoding ?? Encoding.UTF8, detectEncodingFromByteOrderMarks: true, DefaultStreamReaderBufferSize, leaveOpen: true))
+            {
+                CsvLineEnumerator linesEnumerator = new CsvLineEnumerator(streamReader);
```
Comment (pgovind) [inline src/Microsoft.Data.Analysis/DataFrame.IO.cs:376]: Just a couple structs to wrap a `StreamReader` within an `IEnumerable<string>`.

File: src/Microsoft.Data.Analysis/DataFrame.IO.cs [hunk 3]
```diff
@@ -214,87 +193,189 @@ private static DataFrameColumn CreateColumn(Type kind, string[] columnNames, int
...
+        /// <param name="addIndexColumn">add one column with the row index</param>
+        /// <param name="encoding">The character encoding. Defaults to UTF8 if not specified</param>
+        /// <returns><see cref="DataFrame"/></returns>
+        public static DataFrame LoadCsvFromString(string csvString,
+                                char separator = ',', bool header = true,
+                                string[] columnNames = null, Type[] dataTypes = null,
+                                long numberOfRowsToRead = -1, int guessRows = 10, bool addIndexColumn = false,
+                                Encoding encoding = null)
+        {
+            string[] lines = csvString.Split(new[] { Environment.NewLine }, StringSplitOptions.None);
```
Comment (eerhardt) [inline src/Microsoft.Data.Analysis/DataFrame.IO.cs:341]: Are we just completely giving up on quoting? Or is this just a "first effort" to get the API in?

File: src/Microsoft.Data.Analysis/DataFrame.IO.cs [hunk 4]
```diff
@@ -214,87 +193,189 @@ private static DataFrameColumn CreateColumn(Type kind, string[] columnNames, int
...
+        /// <param name="numberOfRowsToRead">number of rows to read not including the header(if present)</param>
+        /// <param name="guessRows">number of rows used to guess types</param>
+        /// <param name="addIndexColumn">add one column with the row index</param>
+        /// <param name="encoding">The character encoding. Defaults to UTF8 if not specified</param>
+        /// <returns><see cref="DataFrame"/></returns>
+        public static DataFrame LoadCsvFromString(string csvString,
+                                char separator = ',', bool header = true,
+                                string[] columnNames = null, Type[] dataTypes = null,
+                                long numberOfRowsToRead = -1, int guessRows = 10, bool addIndexColumn = false,
+                                Encoding encoding = null)
```
Comment (eerhardt) [inline src/Microsoft.Data.Analysis/DataFrame.IO.cs:339]: Encoding doesn't make any sense for a string. strings are always UTF-16.

File: src/Microsoft.Data.Analysis/DataFrame.IO.cs [hunk 5]
```diff
@@ -214,87 +193,189 @@ private static DataFrameColumn CreateColumn(Type kind, string[] columnNames, int
...
-                    columns.Insert(0, indexColumn);
+                    indexColumn[i] = i;
                 }
-                return ret;
+                columns.Insert(0, indexColumn);
+            }
+            return ret;
+        }
+
+        private struct CsvLines : IEnumerable<string>
```
Comment (eerhardt) [inline src/Microsoft.Data.Analysis/DataFrame.IO.cs:275]: Does this need to be a `struct`? You are going to box it as soon as you pass it into `ReadCsvLinesIntoDataFrame(IEnumerable<string> lines` anyway.

If you really wanted to save the single alloc, you could make `ReadCsvLinesIntoDataFrame(IEnumerable<string> lines` be generic `where T : IEnumerable<string>`.

File: src/Microsoft.Data.Analysis/DataFrame.IO.cs [hunk 6]
```diff
@@ -214,87 +193,189 @@ private static DataFrameColumn CreateColumn(Type kind, string[] columnNames, int
...
+                currentLine = null;
+            }
+
+            public string Current => currentLine;
+
+            object IEnumerator.Current => currentLine;
+
+            public void Dispose()
+            {
+                throw new NotImplementedException();
```
Comment (eerhardt) [inline src/Microsoft.Data.Analysis/DataFrame.IO.cs:306]: Why throw on Dispose?

File: src/Microsoft.Data.Analysis/DataFrame.IO.cs [hunk 7]
```diff
@@ -214,87 +193,189 @@ private static DataFrameColumn CreateColumn(Type kind, string[] columnNames, int
...
+            }
+
+            public bool MoveNext()
+            {
+                currentLine = streamReader.ReadLine();
+                return currentLine != null;
+            }
+
+            public void Reset()
```
Comment (eerhardt) [inline src/Microsoft.Data.Analysis/DataFrame.IO.cs:315]: Do we need to support Reset()? Who calls it? You can save a `long` field if it wasn't implemented.

Comment (eerhardt) [reply src/Microsoft.Data.Analysis/DataFrame.IO.cs:315]: Ah, nevermind. I see where we call it now.

File: src/Microsoft.Data.Analysis/DataFrame.IO.cs (hunk 3 above)
Comment (pgovind) [reply src/Microsoft.Data.Analysis/DataFrame.IO.cs:341]: This is just a "first effort". Or more accurately the easiest way to get this API in with just some minor refactoring. I'll work on a proper csv parser shortly and re-visit this implementation at that point. I still think splitting on a new line and sending each line to the csv parser will work.

File: src/Microsoft.Data.Analysis/DataFrame.IO.cs (hunk 4 above)
Comment (pgovind) [reply src/Microsoft.Data.Analysis/DataFrame.IO.cs:339]: Ah yup you're right. Fixed

File: src/Microsoft.Data.Analysis/DataFrame.IO.cs (hunk 5 above)
Comment (pgovind) [reply src/Microsoft.Data.Analysis/DataFrame.IO.cs:275]: Oh, I didn't realize that value type -> interfaces also does boxing. Just looked it up. Nice! Fixed it now. Just made it a class.

File: src/Microsoft.Data.Analysis/DataFrame.IO.cs (hunk 6 above)
Comment (pgovind) [reply src/Microsoft.Data.Analysis/DataFrame.IO.cs:306]: I put that in so we catch it if someone calls this method. We're not dealing with any unmanaged resources here, so no one should be calling `Dispose`

**Review by eerhardt — APPROVED** [review]
LGTM

**Pull request merged**